mysql-test/suite/binlog/t/binlog_recover_async_rollback_trx.test

```diff
@@ -0,0 +1,93 @@
+# active and the background recovery rollback thread will rollback the
+# transactions asynchronously.
+################################################################################
+--source include/have_debug.inc
+--source include/have_debug_sync.inc
+--source include/have_innodb.inc
+--source include/have_binlog_format_row.inc
+
+CREATE TABLE t1 (id int primary key, data int) ENGINE = InnoDB;
+INSERT INTO t1 VALUES (0, 1);
+
+--echo #
+--echo # 1. Check DML in prepared state can rollback correctly.
+--echo #
+
+--echo [ connection con1 ]
+--connect(con1, localhost, root,,)
+SET debug_sync = "ha_commit_trans_after_prepare SIGNAL prepared1 WAIT_FOR continue";
+--send INSERT INTO t1 VALUES(1, 1);
+
+--connect(con2, localhost, root,,)
+SET debug_sync = "ha_commit_trans_after_prepare SIGNAL prepared2 WAIT_FOR continue";
+--send UPDATE t1 SET data = data + 1 WHERE id = 0
+
+--echo [ connection default ]
+--connection default
+SET debug_sync = "now WAIT_FOR prepared1";
+SET debug_sync = "now WAIT_FOR prepared2";
+--source include/kill_mysqld.inc
+
+--disconnect con1
+--disconnect con2
+
+# With the debug option, recovery rollback thread just rolls back the
+# first prepared transaction and then goes to sleep.
+--let $restart_parameters= --debug=d,debug_innodb_rollback_thread --log-error=$MYSQLTEST_VARDIR/tmp/log.err
+--source include/start_mysqld.inc
+--let $wait_condition= SELECT count(*) = 1 FROM information_schema.innodb_trx
+--source include/wait_condition.inc
+
+# Check whether the DML trx has been rolled back asynchronously.
+--let $assert_file= $MYSQLTEST_VARDIR/tmp/log.err
+--let $assert_text= Server reported: recovery_rollback_thread
+--let $assert_select= Rolled back recovered transaction
+--let $assert_count= 1
+--source include/assert_grep.inc
+--remove_file $MYSQLTEST_VARDIR/tmp/log.err
+
+# The previous INSERT is rolled back, so this INSERT will succeed.
+INSERT INTO t1 VALUES(1, 1);
+
+--echo # Expected (0, 1), (1, 1)
+SELECT * FROM t1;
+
+--echo # Expect (0,2), (1,1)
+SET SESSION TRANSACTION ISOLATION LEVEL READ UNCOMMITTED;
+SELECT * FROM t1;
+
+--echo #
+--echo # 2. Check DML can rollback automatically after second crash.
+--echo #
+--source include/kill_mysqld.inc
+--let $restart_parameters=
+--source include/start_mysqld.inc
+--let $wait_condition= SELECT count(*) = 0 FROM information_schema.innodb_trx
+--source include/wait_condition.inc
+
+# Check whether the DML trx has been rolled back asynchronously.
+--let $assert_file= $MYSQLTEST_VARDIR/log/mysqld.1.err
+--let $assert_text= Server reported: Rolled back recovered transaction
+--let $assert_select= Rolled back recovered transaction
+--let $assert_count= 1
+--source include/assert_grep.inc
+
+# The transaction was set to ACTIVE in previous recovery, thus it should not
+# find any prepared thransactions.
+--let $assert_file= $MYSQLTEST_VARDIR/log/mysqld.1.err
+--let $assert_text= Server NOT report: Found 1 prepared transaction
```

Comment:
In 'Server NOT report', is 'NOT' a typo, right? Maybe 'NOW' meant to be...

Reply:
it is NOT. since the transactions has been set to ACTIVE in the first recovery. So it should not find any prepared transaction in the second recovery.